### PR TITLE
Add hold_render to plot

### LINKED ISSUE #12171: [FEATURE] Batch rendering of many glyphs on the same plot
### Problem description

I am experimenting with the following code. We plot 10 lines with their own datasources on one plot. The streaming happens periodically every second for all lines. Each line has a different amount of points generated within this one second.

```
from bokeh.layouts import column
from bokeh.plotting import figure, curdoc
from bokeh.models import ColumnDataSource, Dropdown

import numpy as np
import time

class Line:
	def __init__(self, doc, plot, waiting_time):
		self._current_x = time.monotonic()
		self._current_y = 0
		self._new_points = dict(x=[], y=[])
		self._source = ColumnDataSource(data=dict(x=[], y=[]))

		plot.line(source=self._source)
		doc.add_periodic_callback(self._generate_point, waiting_time)

	def _generate_point(self):
		self._current_x = time.monotonic()
		self._current_y += np.random.normal()

		self._new_points["x"].append(self._current_x)
		self._new_points["y"].append(self._current_y)

	def stream(self):
		self._source.stream(self._new_points)

		self._new_points["x"].clear()
		self._new_points["y"].clear()

doc = curdoc()

p = figure(width=300, height=300, tools="pan,reset,save", y_axis_location="right")
lines = []
for i in range(1, 11):
	lines.append(Line(doc, p, 100*i))

def stream_point():
	for line in lines:
		line.stream()

menu = [("Item 1", "item_1"), ("Item 2", "item_2"), ("Item 3", "item_3")]
dropdown = Dropdown(label="Dropdown button", menu=menu)

doc.add_periodic_callback(stream_point, 1000)
doc.add_root(column(p, dropdown))
```

The problem here is that after some time the browser spends so much time rendering the plot that the dropdown button becomes unusable.

### Feature description

Taking a look at the profiling tools that Firefox offers I can spot that for every time the plot receives a websocket message it redraws the whole plot. You can see on the picture the black line (websocket message) followed by a brown line (rendering).

image

What would be really great and would solve my problem here is to trigger rendering only after the last websocket message. Since my update loop is self contained I think the simplest solution would be to give me a possibility to trigger the rendering myself.

Like so
```
plot(automatic_redraw=False)

def stream_point():
	for line in lines:
		line.stream()
        plot.redraw()
```
or
```
def stream_point():
        plot.render_on_update =False
	for line in lines:
                if is_last_line():
                        plot.render_on_update =True
		line.stream()
```

### Potential alternatives

Another idea that comes to my mind is to set some kind of timer on the first received stream message and don't trigger rendering. While the timer is running all next updates are processed but also don't trigger rendering. Once the timer goes it rerenders the plot. On the next received stream message we start from the beginning.

### Additional information

I am happy to look into implementing one of those myself since ideally it should be a simple change. Not sure though if those proposals don't break some design principle of bokeh.

### PR DESCRIPTION
**Pull request by Kadek**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [X] issues: fixes #12171
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

I have based this on branch-2.4 so that I can use the fork locally straight away.
I tried branching from branch-3.0 but I couldn't get the bokehjs to build on my Windows machine.

I am not sure whether this should be covered by unit tests or additional documentation. Happy to add it if you point me to specific area.

### CODE REVIEW

**Comment by bryevdv** [comment]
@Kadek we don’t intend to have another 2.x release, so any PR would need to be against `branch-3.0`. 

cc @bokeh/core for thoughts about the approach in general.

**Comment by mattpap** [comment]
> I tried branching from branch-3.0 but I couldn't get the bokehjs to build on my Windows machine.

Any specifics? bokehjs should build on windows just fine, unless we broke something recently. We don't use windows very often, and usually only test there before releases.

File: bokeh/models/plots.py [hunk 1]
```diff
@@ -756,6 +756,29 @@ def _scale(cls, scale: Literal["auto", "linear", "log", "categorical"]) -> Scale
...
+    The problem with this code is that every stream triggers a re-rendering of the plot.
+    Even tough repainting only on the last stream would produce almost identical visual effect.
+    Especially for lines with many points this becomes computationally expensive and can freeze your browser.
+    Using hold_render we can control when rendering is initiated like so:
+
+    plot.hold_render = True
+    for line in lines:
+        if is_last(line):
+            plot.hold_render = False
+        line.stream(new_points())
```
Comment (mattpap) [inline bokeh/models/plots.py:777]:
```suggestion
    .. code:: python
    
        plot.hold_render = True
        for line in lines:
            if is_last(line):
                plot.hold_render = False
            line.stream(new_points())
```

Comment (mattpap) [reply bokeh/models/plots.py:777]: Same above.

**Comment by mattpap** [comment]
This:
```
    plot.hold_render = True
    for line in lines:
        if is_last(line):
            plot.hold_render = False
        line.stream(new_points())
```
will be really cumbersome to work with and inconsistent with how similar things work elsewhere in bokeh. The way this code should look like is this:
```
    plot.hold_render = True
    for line in lines:
        line.stream(new_points())
    plot.hold_render = False
```
or even better with some context manager:
```
    with plot.hold(render=True):
        for line in lines:
            line.stream(new_points())
```
(artist's impression).

To allow this, in place of the current implementation there should be an `on_change()` for `hold_render`, which would `pause()` and `unpause()` respectively, given the value of the property. This way, setting `plot.hold_render = False` will force paint if needed.

**Comment by Kadek** [comment]
Oh I love the context manager idea. Do you have some example already existing in bokeh that I could use as a reference for code quality?

**Comment by mattpap** [comment]
> Do you have some example already existing in bokeh that I could use as a reference for code quality?

There are a few across the repository; grep for `contextmanager`. The pattern (give or take) is as follows:
```py
from contextlib import contextmanager

class Plot:
    ...
    @contextmanager
    def hold(self):
         self.hold_render = True
         yield
         self.hold_render = False
```

**Comment by Kadek** [comment]
I managed to get it to build on bokeh-3.0. I implemented the contextmanager change there as well
https://github.com/bokeh/bokeh/pull/12236

**Comment by mattpap** [comment]
It would be possible to reuse this PR, by rebasing the branch with `git rebase --onto origin/branch-3.0 HEAD~1` (in this particular case), force pushing and changing the base branch of this PR. Though starting a new PR is a viable and perhaps easier option.

**Pull request closed without merging**